### ISSUE TITLE
Life cycle policy interferes with podbird functionality 

### ISSUE THREAD

**Issue by userj** [ubports/ubuntu-touch#312]
- Device: Bacon
- Channel:Bacon devel
- Build: r185

## Steps to reproduce

1. Open podbird
2. Play audio episode
3. Move podbird to background
4. Close podbird
5. open podbird
6. Play episode again

## Expected behavoir
Episode picks up where left off
...

## Actual behavior
Episode starts from beginning
...

## Logfiles and additional information
https://bugs.launchpad.net/bugs/1725377

**Comment by ChloeWolfieGirl** [comment]
Just gonna bring  Michaels comment here!
"It's a limitation of the life cycle policy, there's nothing Podbird can do when suspended so there's no way of working around it (other than exempting Podbird from the life cycle management completely). Ideally I think apps should be woken up and sent a termination signal when the user closes them to allow them to save state."

Apps shouldn't be completely removed from the life cycle management, as it will hurt your battery life, a lot!

**Comment by NeoTheThird** [comment]
@userj Thanks for reporting, this has been a known but untracked issue for a while now. Also see: #194.

IIRC @sverzegnassi and @dpniel had some ideas on how to handle this issue, could you share them here?

**Comment by startcloud** [comment]
Some reference to old discussions:
https://lists.launchpad.net/ubuntu-phone/msg03748.html
http://nikwen.github.io/ubuntu/2015/09/17/my-work-on-email-notifications.html
https://www.youtube.com/watch?v=v5iY2NgaBeY

**Comment by sverzegnassi** [comment]
Here on my BQ E5, development channel (r40), everything works as expected.

Podbird resume the episode from the very last second I've been listening to.
For instance: the app is suspended at 05:56; I kill the app when the podcast is at 06:10; next time I launch the app, audio is resumed from 06:10.

I also tried a small counter app for testing, which stored the value of the counter in a QML Settings backend when the `Component.onDestruction` QML signal is emitted.

When the application is not the top-most, focused window, it gets suspended as usual.
When I close the app from the windows spread, the app is temporary resumed and can perform all the required operations before being terminated (i.e. storing the value of the counter in a QSettings backend). Next time I launch the app, the previous value is properly restored.

There used to be a bad implementation of the kill signal in old versions of Mir/Unity 8: when an app gets closed from the windows spread, a SIGKILL signal is sent, instead of SIGTERM.

@NeoTheThird I suspect when you started to build QtMir from our GitHub repo, you automagically imported the bugfix at:
https://github.com/ubports/qtmir/commit/d0a408dd7e36090e038cbbbdab742fe5c188d7a8


@userj Could you please specify which version of Ubuntu Touch are you running?

**Comment by startcloud** [comment]
@sverzegnassi maybe it has an issue in signals that is fixed in recent versions but here we want to track/discuss the life cycle issue policy that impacts the app devs that's why @NeoTheThird rename it

**Comment by sverzegnassi** [comment]
@startcloud 
First three messages are about a specific issue which has been reported for a specific application, but it could really affect many others - one for example is QuickMemo, available on the Ubuntu Store.
What Jan did makes sense.

My feeling here - in any case - is that the issue has been wrongly addressed to a life cycle restriction, while it might appear to be the result of a non-standard implementation of kill signals.
That shifted the topic of the discussion.

If someone can please confirm that Podbird behaves properly on latest images, so that we can triage the report, and eventually assign it to the OTA-3 milestone.

In that case, no, we wouldn't want to discuss life cycle policies here. We still want to track bugs reported by users, and these would end up to be two different discussions, with two different targets (vivid vs. xenial+halium) - IMHO

======

Anyway, replying to the question: yes, life cycle policies might represent a problem for specific types of applications.

The platform already provides a few services that cover many of the most common use cases: `account-polld`, `media-hub` or the `ubuntu-download-manager`.
However, there are still uncovered cases that prevent development of relevant apps:
- Uploading/sending a content on internet (e.g. telegram-app)
- Downloading files that are not supported by the Ubuntu downloader (e.g. torrent)
- Reproduce audio from an unsupported and/or custom provider (e.g. Spotify).

All these cases would require applications to implement their custom handler.

However, I suspect this is just a subset of a bigger problem.

Most of the UT middleware have been rewritten from scratch in a non-standard way and/or is poorly documented. Without the traction of the main consumer GNU/Linux distribution (Ubuntu Desktop), or phones being distributed by large retailers, it nearly impossible to see developers supporting the platform. Also, support for many APIs has been provided only for QML-based apps (e.g. ContentHub).

At the moment - whatever the situation is - we first need to prove that apps can run in background without high battery consumption. That could affect any further decision.

I've been testing a lifecycle exception for telegram-app for a while, and I didn't noticed any strange consumption. My hope is that apps like Telegram could just be launched as daemon at device startup, like it happens on SailfishOS/Maemo/Meego, and - more generally - on desktop.

Another relevant question for the final decision depends on how we want to provide convergence, and whether we want to support Snap'd apps or not.
I suspect that recent changes in Mir/MirAL and snapd might force us to design our platform differently.

Canonical is being famous these days for deprecating existing functionalities, so I wouldn't be surprise if the Mir client integration will be deprecated and remove, once the Wayland client support will be finished. Also we'd need to know which is the plan with the `unity8` interface in snapd, which appear to be de-facto deprecated.

This might force us to design our platform exactly as a desktop (e.g. drop any lifecycle restriction), in order to be fully aligned with upstream decisions (and avoid unwanted surprises).
Anyway, this falls outside of my duties, and I trust any decision Marius and others will make.

We could even decide to stay with the current Click packaging format, and keep certain restrictions alive (e.g. the `account-polld` thing).

Just my debatable opinion - I might be wrong with some argument.

My hard guess is that we ideally want to allow app devs to write their app once with e.g. QtQuick Controls 2, and allow them to distribute a single Snap package that runs on Ubuntu Touch, Ubuntu, Solus, etc. without further changes. Dekko2 experience is very meaningful, in this regard.

**Comment by Flohack74** [comment]
When I read this from the beginning, I can only think about Podbird being killed by OOM killer, thats the only situation where it could loose the current state. Of course, playing in the background would be very nice. There are ways to enable lifecycle exceptions, but its not easy for end users to tweak that. In fact, sometimes it would even save more battery if the lifecycle would not be in place, like Telegram:

- Every time the app goes to the background, you risk loosing the socket connection and next time you need a full initial handshake
- Catching up with updates after hours instead of seconds/minutes will need more protocol overhead

In summary while it seems logical that no background app saves battery, it could be the opposite way in some special cases.

BR

**Comment by startcloud** [comment]
@sverzegnassi you are right, I responded quickly since I am attracted with the name of this issue and ignored the kill signal issue so to sum up:
- We
 should clarify if Podbird behaves properly on latest images if confirmed we should split this topic in two separate IDs.
- We agree that we should do something for the strict life cycle policy, may be this is not the right place to discuss it but I don't find any tracking of if on Ubports project (only few questions in community updates)  so it is good to keep and open point here.
- We could make some measurement about battery consumption when applications start in background to base our analysis on real values and not suppositions.

I am ready to help on that but I don't know how can I enable this life cycle exception, could someone send me instruction to enable it for one or several applications?
( I have several remarks on comment 5 but I want to progress more than debate in some philosophical theories)

**Comment by dobey** [comment]
There's an important piece of information that seems to be missing here, particularly for the originally mentioned Podbird case:

When an application is stopped by the lifecycle policy, it is given a signal and has a couple seconds to save state. To save state, all one has to do is handle the signal, and then restore it when the app is started again.

**Comment by dobey** [comment]
Further, I don't think removing the lifecycle policy is at all a good idea. It's a very core part of how apps are launched and managed under Unity8, and it would probably be easier to write most of unity8 from scratch than to start pulling such core pieces of functionality from the system.

Yes, there are special cases where it inhibits development of certain things. I had been very vocal about it in the past, as well, as it inhibits certain types of apps which I wanted to build. However, removing security is not a solution to the problem of security causing it to be more difficult to implement some application.

**Comment by PhoenixLandPirate** [comment]
I must agree with @dobey there are definitely issues, a big one we're fighting with currently would be notifications, telegram, text, clock, and some other apps have good notifications, but you cant get them from uMatriks, signal, etc.

I think the question is, what do you think success is?
I think success is being more secure and private then the alternatives.
Which is more important then getting a ton more users.

**Comment by dobey** [comment]
I think notifications issues are irrelevant here. One can implement push notifications on server side (doable for matrix), or add an additional tool/script to be run by accounts-polld to poll for new notifications in background. Polling is of course not good for things like [Signal](https://github.com/signalapp/Signal-Server), but it is open source, so perhaps it would be possible to get push notification support there too.

I think if we can concentrate more on the hard problems first, we can deal with less important problems like how to get more native apps, later. When 16.04 is stable with anbox, now many people are going to just be running Android apps there anyway, versus native apps? It might be a good stepping stone, but I think it's something we want to move back away from in the future.

**Comment by userj** [comment]
Some related updates to the original report here: https://bugs.launchpad.net/podbird/+bug/1725377

LIfe cycle affects database interaction? or is database a different issue?

**Comment by Flohack74** [comment]
If it is on close then the App misses a shutdown handler, or needs to store the data earlier. A close of an App is very brute-forced and may come unexpected, any time. Data in memory will be lost if the App does not store it properly.

**Comment by PhoenixLandPirate** [comment]
Maybe we could make it less "brute-forced" instead of killing the app as soon as you fling the app away, perhaps we could implement something like Micheal suggested.

Perhaps you could fling the app away, and visually it closes, but it actually activates and becomes awake for a few seconds, just to tell the app that its being closed, and then it gives the app a breath o update and store the data it needs to?

**Comment by dobey** [comment]
> Perhaps you could fling the app away, and visually it closes, but it actually activates and becomes awake for a few seconds, just to tell the app that its being closed, and then it gives the app a breath o update and store the data it needs to?

This is already done, prior to the app being sent a SIGTERM. But if apps aren't saving state properly, then you will lose it, regardless of any changes made here. Apps really should be maintaining state whenever a user action occurs, not when a theoretical pause of the app might happen.
